keps/sig-cli/kustomize-exec-secret-generator.md

```diff
@@ -0,0 +1,195 @@
+
+- Alice wants to develop an Application requiring a shared Secret, and to deploy it on Kubernetes using gitops
+- Alice wants her gitops deployment mechanism to pull the Secrets that it deploys from an
+  remote source without writing the Secrets as files to local disk.
+- Alice's organization configures the gitops deployment container to run kustomize in the cluster
+  and be capable of pulling Secrets from remote locations
+- Alice writes her kustomization.yaml to use the generation options configured by her organization.
+
+Example exploit to avoid:
+
+- Alice wants to run a whitebox mysql instance on a test cluster
+- Carl publishes a whitebox mysql `kustomization.yaml` on GitHub, with a SecretGenerator
+  that will read Alice's ~/.kube/config and send it to Carl's server by executing `sh`
+  will a script to generate some Secret
+- Alice runs `kubectl apply -k https://github.com/carl/mysql` and has the credentials
+  of all of her Kubernetes clusters sent to Carl when the Secret is generated.
+
+See [692][execRemoval] for more details.
+
+## Motivation
+
+The ability to create Kubernetes Secrets generated by commands is a commonly requested user.
+This is useful for use cases where the user doesn't want the Secrets to be appear decrypted
+on disk before being applied to the cluster.
+
+Examples:
+
+- Secrets sources are encrypted and need to be decrypted before they are applied
+- Secrets sources are not stored locally, and need to be fetched from some remote location
+- Secrets are generated using some function (e.g. private-keys)
+
+**Note:** For the common targeted case, *the solution for generating the Secret already
+exists as a command on the user's machine*.  User's are not expected to require a
+marketplace of solutions, rather instead they are expected to need to be able to
+invoke the tools they already depend upon.
+
+### Goals
+
+- Enable users to generate Secrets using the tools they already have on their machines
+- Require Alice to opt-in inorder for Carl to execute commands on her machine
+- Support Linux / Mac / Windows platforms
+
+### Non-Goals
+
+- Enable an ecosystem of users authoring plugins and sharing them
+- Reduce friction for publicly published whitebox `kustomization.yaml`s to generate Secrets
+
+## Proposal
+
+Re-introduce `exec` Secret generators, but require the executables to have a whitelisted prefix
+and require a flag with the search location for the prefixes.
+
+- Introduce a flag `--sg-exec-prefix` defaulted to `kustomize-sg-`
+  - if set to the empty string, no prefix is used
+- Introduce a flag `--sg-exec-path` defaulted to ``
+  - Defines where to find the executables
+  - May be one of:
+    - (default) the empty string - disable exec secrets altogether
+    - `$PATH` - any executable in the user's path matching the prefix
+    - an absolute directory path - only executables matching the prefix in that directory
+    - a relative directory - only executables matching the prefix in that directory
```

Comment:
I'm assuming it's relative to the `kustomization.yaml` file that was command-line specified, and by "in that directory" you're saying "that doesn't contain ..". Does it follow symlinks outside "that directory"?

Reply:
changed this.  lmk what you think